File: python/selfie-lib/selfie_lib/CommentTracker.py
import threading
from collections.abc import Iterable
from enum import Enum, auto

from .Slice import Slice
from .TypedPath import TypedPath
from .WriteTracker import CallStack, SnapshotFileLayout


class WritableComment(Enum):
    NO_COMMENT = auto()
    ONCE = auto()
    FOREVER = auto()

    @property
    def writable(self) -> bool:
        return self != WritableComment.NO_COMMENT


class CommentTracker:
    def __init__(self):
        self.cache: dict[TypedPath, WritableComment] = {}
        self.lock = threading.Lock()

    def paths_with_once(self) -> Iterable[TypedPath]:
        with self.lock:
            return [
                path
                for path, comment in self.cache.items()
                if comment == WritableComment.ONCE
            ]

    def hasWritableComment(self, call: CallStack, layout: SnapshotFileLayout) -> bool:
        path = layout.sourcefile_for_call(call.location)
        with self.lock:
            if path in self.cache:
                return self.cache[path].writable
            else:
                new_comment, _ = self.__commentAndLine(path)
                self.cache[path] = new_comment
                return new_comment.writable

    @staticmethod
    def commentString(typedPath: TypedPath) -> tuple[str, int]:
        comment, line = CommentTracker.__commentAndLine(typedPath)
        if comment == WritableComment.NO_COMMENT:
            raise ValueError("No writable comment found")
        elif comment == WritableComment.ONCE:
            return ("#selfieonce", line)
        elif comment == WritableComment.FOREVER:
            return ("#SELFIEWRITE", line)
        else:
            raise ValueError("Invalid comment type")

    @staticmethod
    def __commentAndLine(typedPath: TypedPath) -> tuple[WritableComment, int]:
        with open(typedPath.absolute_path, encoding="utf-8") as file:
            content = Slice(file.read())
        for comment_str in [
            "# selfieonce",
            "#selfieonce",
            "# SELFIEWRITE",
            "#SELFIEWRITE",
        ]:
            index = content.indexOf(comment_str)
            if index != -1:
                lineNumber = content.baseLineAtOffset(index)
                comment = (
                    WritableComment.ONCE
                    if "once" in comment_str
                    else WritableComment.FOREVER
                )
                return (comment, lineNumber)
        return (WritableComment.NO_COMMENT, -1)

File: jvm/selfie-lib/src/commonMain/kotlin/com/diffplug/selfie/guts/CommentTracker.kt
/*
 * Copyright (C) 2024-2025 DiffPlug
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.diffplug.selfie.guts

import com.diffplug.selfie.ArrayMap

/**
 * Tracks whether a given file has a comment which allows it to be written to. Thread-safe on
 * multithreaded platforms.
 */
class CommentTracker {
  private enum class WritableComment {
    NO_COMMENT,
    ONCE,
    FOREVER;
    val writable: Boolean
      get() = this != NO_COMMENT
  }
  private val cache = atomic(ArrayMap.empty<TypedPath, WritableComment>())
  fun pathsWithOnce(): Iterable<TypedPath> =
      cache.get().mapNotNull { if (it.value == WritableComment.ONCE) it.key else null }
  fun hasWritableComment(call: CallStack, layout: SnapshotFileLayout): Boolean {
    val path = layout.sourcePathForCall(call.location)
    val comment = cache.get()[path]
    return if (comment != null) {
      comment.writable
    } else {
      val newComment = commentAndLine(path, layout.fs).first
      // may race get(), ignore if already present
      cache.updateAndGet { it.plusOrNoOp(path, newComment) }
      newComment.writable
    }
  }

  companion object {
    fun commentString(typedPath: TypedPath, fs: FS): Pair<String, Int> {
      val (comment, line) = commentAndLine(typedPath, fs)
      return when (comment) {
        WritableComment.NO_COMMENT -> throw UnsupportedOperationException()
        WritableComment.ONCE -> Pair("//selfieonce", line)
        WritableComment.FOREVER -> Pair("//SELFIEWRITE", line)
      }
    }
    private fun commentAndLine(typedPath: TypedPath, fs: FS): Pair<WritableComment, Int> {
      // TODO: there is a bug here due to string constants, and non-C file comments
      val content = Slice(fs.fileRead(typedPath))
      for (comment in listOf("//selfieonce", "// selfieonce", "//SELFIEWRITE", "// SELFIEWRITE")) {
        val index = content.indexOf(comment)
        if (index != -1) {
          val lineNumber = content.baseLineAtOffset(index)
          val comment =
              if (comment.contains("once")) WritableComment.ONCE else WritableComment.FOREVER
          return Pair(comment, lineNumber)
        }
      }
      return Pair(WritableComment.NO_COMMENT, -1)
    }
  }
}
